File: tests/wasm/test_apo.js

import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import {
    loadTestData,
    assertArrayClose,
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {

    try {
        const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
        const importPath = process.platform === 'win32'
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);

    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }

    testData = loadTestData();
});

test('APO partial params', () => {

    const close = new Float64Array(testData.close);

    const result = wasm.apo_js(close, 10, 20);
    assert.strictEqual(result.length, close.length);
});

test('APO accuracy', async () => {

    const close = new Float64Array(testData.close);
    const expected = EXPECTED_OUTPUTS.apo;

    const result = wasm.apo_js(
        close,
        expected.defaultParams.short_period,
        expected.defaultParams.long_period
    );

    assert.strictEqual(result.length, close.length);


    const last5 = result.slice(-5);
    assertArrayClose(
        last5,
        expected.last5Values,
        1e-1,
        "APO last 5 values mismatch"
    );


    await compareWithRust('apo', result, 'single', expected.defaultParams);
});

test('APO default params', () => {

    const close = new Float64Array(testData.close);

    const result = wasm.apo_js(close, 10, 20);
    assert.strictEqual(result.length, close.length);
});

test('APO zero period', () => {

    const data = new Float64Array([10.0, 20.0, 30.0]);

    assert.throws(() => {
        wasm.apo_js(data, 0, 20);
    }, /Invalid period/);
});

test('APO period invalid', () => {

    const data = new Float64Array([10.0, 20.0, 30.0]);

    assert.throws(() => {
        wasm.apo_js(data, 20, 10);
    }, /short_period not less than long_period/);

    assert.throws(() => {
        wasm.apo_js(data, 10, 10);
    }, /short_period not less than long_period/);
});

test('APO very small dataset', () => {

    const single_point = new Float64Array([42.0]);

    assert.throws(() => {
        wasm.apo_js(single_point, 9, 10);
    }, /Not enough valid data/);
});

test('APO reinput', () => {

    const close = new Float64Array(testData.close);


    const firstResult = wasm.apo_js(close, 10, 20);
    assert.strictEqual(firstResult.length, close.length);


    const secondResult = wasm.apo_js(
        new Float64Array(firstResult),
        5,
        15
    );
    assert.strictEqual(secondResult.length, firstResult.length);
});

test('APO NaN handling', () => {

    const close = new Float64Array(testData.close);

    const result = wasm.apo_js(close, 10, 20);
    assert.strictEqual(result.length, close.length);


    if (result.length > 30) {
        for (let i = 30; i < result.length; i++) {
            assert(!isNaN(result[i]), `Found unexpected NaN at index ${i}`);
        }
    }
});

test('APO all NaN input', () => {

    const allNaN = new Float64Array(100);
    allNaN.fill(NaN);

    assert.throws(() => {
        wasm.apo_js(allNaN, 10, 20);
    }, /All values are NaN/);
});

test('APO batch single parameter set', () => {

    const close = new Float64Array(testData.close);


    const batchResult = wasm.apo_batch_js(
        close,
        10, 10, 0,
        20, 20, 0
    );


    const singleResult = wasm.apo_js(close, 10, 20);

    assert.strictEqual(batchResult.length, singleResult.length);
    assertArrayClose(batchResult, singleResult, 1e-10, "Batch vs single mismatch");
});

test('APO batch multiple parameters', () => {

    const close = new Float64Array(testData.close.slice(0, 100));


    const batchResult = wasm.apo_batch_js(
        close,
        5, 15, 5,
        20, 30, 10
    );


    const expectedCombos = 6;
    assert.strictEqual(batchResult.length, expectedCombos * 100);


    const firstRow = batchResult.slice(0, 100);
    const singleResult = wasm.apo_js(close, 5, 20);
    assertArrayClose(firstRow, singleResult, 1e-10, "First combination mismatch");
});

test('APO batch metadata', () => {

    const metadata = wasm.apo_batch_metadata_js(
        5, 15, 5,
        20, 30, 10
    );


    assert.strictEqual(metadata.length, 12);


    const expected = [
        5, 20,
        5, 30,
        10, 20,
        10, 30,
        15, 20,
        15, 30
    ];

    for (let i = 0; i < expected.length; i++) {
        assert.strictEqual(metadata[i], expected[i], `Metadata mismatch at index ${i}`);
    }
});

test('APO batch invalid combinations', () => {

    const close = new Float64Array(testData.close.slice(0, 50));


    const batchResult = wasm.apo_batch_js(
        close,
        10, 30, 10,
        15, 25, 10
    );


    assert.strictEqual(batchResult.length, 3 * 50);
});

test('APO batch empty result', () => {

    const close = new Float64Array([1, 2, 3, 4, 5]);


    assert.throws(() => {
        wasm.apo_batch_js(
            close,
            20, 30, 10,
            10, 15, 5
        );
    }, /Invalid period|invalid range/);
});


test('APO batch - new ergonomic API with single parameter', () => {

    const close = new Float64Array(testData.close);

    const result = wasm.apo_batch(close, {
        short_period_range: [10, 10, 0],
        long_period_range: [20, 20, 0]
    });


    assert(result.values, 'Should have values array');
    assert(result.combos, 'Should have combos array');
    assert(typeof result.rows === 'number', 'Should have rows count');
    assert(typeof result.cols === 'number', 'Should have cols count');


    assert.strictEqual(result.rows, 1);
    assert.strictEqual(result.cols, close.length);
    assert.strictEqual(result.combos.length, 1);
    assert.strictEqual(result.values.length, close.length);


    const combo = result.combos[0];
    assert.strictEqual(combo.short_period, 10);
    assert.strictEqual(combo.long_period, 20);


    const oldResult = wasm.apo_js(close, 10, 20);
    for (let i = 0; i < oldResult.length; i++) {
        if (isNaN(oldResult[i]) && isNaN(result.values[i])) {
            continue;
        }
        assert(Math.abs(oldResult[i] - result.values[i]) < 1e-10,
               `Value mismatch at index ${i}`);
    }
});

test('APO batch - new API with multiple parameters', () => {

    const close = new Float64Array(testData.close.slice(0, 50));

    const result = wasm.apo_batch(close, {
        short_period_range: [5, 15, 5],
        long_period_range: [20, 30, 10]
    });


    assert.strictEqual(result.rows, 6);
    assert.strictEqual(result.cols, 50);
    assert.strictEqual(result.combos.length, 6);
    assert.strictEqual(result.values.length, 300);


    const expectedCombos = [
        [5, 20], [5, 30],
        [10, 20], [10, 30],
        [15, 20], [15, 30]
    ];

    for (let i = 0; i < expectedCombos.length; i++) {
        assert.strictEqual(result.combos[i].short_period, expectedCombos[i][0]);
        assert.strictEqual(result.combos[i].long_period, expectedCombos[i][1]);
    }


    const secondRow = result.values.slice(result.cols, 2 * result.cols);
    assert.strictEqual(secondRow.length, 50);


    const oldResult = wasm.apo_js(close, 5, 20);
    const firstRow = result.values.slice(0, result.cols);
    for (let i = 0; i < oldResult.length; i++) {
        if (isNaN(oldResult[i]) && isNaN(firstRow[i])) {
            continue;
        }
        assert(Math.abs(oldResult[i] - firstRow[i]) < 1e-10,
               `Value mismatch at index ${i}`);
    }
});

test('APO batch - new API matches old API results', () => {

    const close = new Float64Array(testData.close.slice(0, 100));

    const params = {
        short_period_range: [5, 15, 5],
        long_period_range: [20, 30, 5]
    };


    const oldValues = wasm.apo_batch_js(
        close,
        params.short_period_range[0], params.short_period_range[1], params.short_period_range[2],
        params.long_period_range[0], params.long_period_range[1], params.long_period_range[2]
    );


    const newResult = wasm.apo_batch(close, params);


    assert.strictEqual(oldValues.length, newResult.values.length);

    for (let i = 0; i < oldValues.length; i++) {
        if (isNaN(oldValues[i]) && isNaN(newResult.values[i])) {
            continue;
        }
        assert(Math.abs(oldValues[i] - newResult.values[i]) < 1e-10,
               `Value mismatch at index ${i}: old=${oldValues[i]}, new=${newResult.values[i]}`);
    }
});

test('APO batch - new API error handling', () => {
    const close = new Float64Array(testData.close.slice(0, 10));


    assert.throws(() => {
        wasm.apo_batch(close, {
            short_period_range: [10, 10]
        });
    }, /Invalid config/);


    assert.throws(() => {
        wasm.apo_batch(close, {
            short_period_range: [5, 15, 5]

        });
    }, /Invalid config/);


    assert.throws(() => {
        wasm.apo_batch(close, {
            short_period_range: "invalid",
            long_period_range: [20, 30, 5]
        });
    }, /Invalid config/);
});

test('APO edge cases', () => {

    const minData = new Float64Array(20).fill(1.0);
    const result = wasm.apo_js(minData, 10, 20);
    assert.strictEqual(result.length, minData.length);


    assert.strictEqual(result[0], 0.0);


    assert(Math.abs(result[result.length - 1]) < 1e-10, "APO should converge to 0 with constant prices");
});


test('APO fast API - basic operation', () => {
    const data = new Float64Array(testData.close);
    const len = data.length;


    const inPtr = wasm.apo_alloc(len);
    const outPtr = wasm.apo_alloc(len);
    assert(inPtr !== 0, 'Failed to allocate input memory');
    assert(outPtr !== 0, 'Failed to allocate output memory');

    try {

        const inView = new Float64Array(wasm.__wasm.memory.buffer, inPtr, len);
        inView.set(data);


        wasm.apo_into(inPtr, outPtr, len, 10, 20);


        const resultView = new Float64Array(wasm.__wasm.memory.buffer, outPtr, len);
        const result = Array.from(resultView);


        const safeResult = wasm.apo_js(data, 10, 20);
        assertArrayClose(result, safeResult, 1e-10, "Fast API vs safe API mismatch");
    } finally {

        wasm.apo_free(inPtr, len);
        wasm.apo_free(outPtr, len);
    }
});

test('APO fast API - in-place operation (aliasing)', () => {
    const data = new Float64Array(testData.close);
    const len = data.length;


    const originalData = new Float64Array(data);


    const ptr = wasm.apo_alloc(len);
    assert(ptr !== 0, 'Failed to allocate memory');

    try {

        const wasmData = new Float64Array(wasm.__wasm.memory.buffer, ptr, len);
        wasmData.set(data);


        wasm.apo_into(ptr, ptr, len, 10, 20);


        const resultView = new Float64Array(wasm.__wasm.memory.buffer, ptr, len);
        const result = Array.from(resultView);


        const expectedResult = wasm.apo_js(originalData, 10, 20);
        assertArrayClose(result, expectedResult, 1e-10, "In-place operation mismatch");
    } finally {
        wasm.apo_free(ptr, len);
    }
});

test('APO fast API - error handling', () => {
    const data = new Float64Array([1, 2, 3, 4, 5]);
    const len = data.length;


    assert.throws(() => {
        wasm.apo_into(0, 0, len, 10, 20);
    }, /Null pointer/);


    const inPtr = wasm.apo_alloc(len);
    const outPtr = wasm.apo_alloc(len);

    try {

        const inView = new Float64Array(wasm.__wasm.memory.buffer, inPtr, len);
        inView.set(data);

        assert.throws(() => {
            wasm.apo_into(inPtr, outPtr, len, 0, 20);
        }, /Invalid period/);

        assert.throws(() => {
            wasm.apo_into(inPtr, outPtr, len, 20, 10);
        }, /short_period not less than long_period/);
    } finally {
        wasm.apo_free(inPtr, len);
        wasm.apo_free(outPtr, len);
    }
});

test('APO fast API - memory management', () => {
    const sizes = [100, 1000, 10000];

    sizes.forEach(size => {
        const ptr = wasm.apo_alloc(size);
        assert(ptr !== 0, `Failed to allocate ${size} elements`);


        const mem = new Float64Array(wasm.__wasm.memory.buffer, ptr, size);
        mem[0] = 42.0;
        mem[size - 1] = 99.0;
        assert.strictEqual(mem[0], 42.0);
        assert.strictEqual(mem[size - 1], 99.0);


        wasm.apo_free(ptr, size);
    });
});

test.after(() => {
    console.log('APO WASM tests completed');
});

File: tests/python/test_apo.py
"""
Python binding tests for APO indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:

    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS
from rust_comparison import compare_with_rust


class TestApo:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()

    def test_apo_partial_params(self, test_data):
        """Test APO with partial parameters - mirrors check_apo_partial_params"""
        close = test_data['close']


        result = ta_indicators.apo(close, 10, 20)
        assert len(result) == len(close)

    def test_apo_accuracy(self, test_data):
        """Test APO matches expected values from Rust tests - mirrors check_apo_accuracy"""
        close = test_data['close']
        expected = EXPECTED_OUTPUTS['apo']

        result = ta_indicators.apo(
            close,
            short_period=expected['default_params']['short_period'],
            long_period=expected['default_params']['long_period']
        )

        assert len(result) == len(close)


        assert_close(
            result[-5:],
            expected['last_5_values'],
            rtol=0,
            atol=1e-1,
            msg="APO last 5 values mismatch"
        )


        compare_with_rust('apo', result)

    def test_apo_default_params(self, test_data):
        """Test APO with default parameters"""
        close = test_data['close']

        result = ta_indicators.apo(close)
        assert len(result) == len(close)

    def test_apo_zero_period(self):
        """Test APO fails with zero period - mirrors check_apo_zero_period"""
        data = np.array([10.0, 20.0, 30.0])

        with pytest.raises(ValueError):
            ta_indicators.apo(data, short_period=0, long_period=20)

    def test_apo_period_invalid(self):
        """Test APO fails when short_period >= long_period - mirrors check_apo_period_invalid"""
        data = np.array([10.0, 20.0, 30.0])

        with pytest.raises(ValueError):
            ta_indicators.apo(data, short_period=20, long_period=10)

        with pytest.raises(ValueError):
            ta_indicators.apo(data, short_period=10, long_period=10)

    def test_apo_very_small_dataset(self):
        """Test APO fails with insufficient data - mirrors check_apo_very_small_dataset"""
        single_point = np.array([42.0])

        with pytest.raises(ValueError):
            ta_indicators.apo(single_point, short_period=9, long_period=10)

    def test_apo_reinput(self, test_data):
        """Test APO applied twice (re-input) - mirrors check_apo_reinput"""
        close = test_data['close']


        first_result = ta_indicators.apo(close, short_period=10, long_period=20)
        assert len(first_result) == len(close)


        second_result = ta_indicators.apo(first_result, short_period=5, long_period=15)
        assert len(second_result) == len(first_result)

    def test_apo_nan_handling(self, test_data):
        """Test APO handles NaN values correctly - mirrors check_apo_nan_handling"""
        close = test_data['close']

        result = ta_indicators.apo(close, short_period=10, long_period=20)
        assert len(result) == len(close)


        if len(result) > 30:
            assert not np.any(np.isnan(result[30:])), "Found unexpected NaN after index 30"

    def test_apo_all_nan_input(self):
        """Test APO with all NaN values"""
        all_nan = np.full(100, np.nan)

        with pytest.raises(ValueError):
            ta_indicators.apo(all_nan)

    def test_apo_kernel_selection(self, test_data):
        """Test APO with different kernel selections"""
        close = test_data['close']


        result_auto = ta_indicators.apo(close, kernel=None)
        result_scalar = ta_indicators.apo(close, kernel='scalar')

        assert len(result_auto) == len(close)
        assert len(result_scalar) == len(close)


        assert_close(result_auto, result_scalar, rtol=1e-10)

    def test_apo_streaming(self, test_data):
        """Test APO streaming functionality"""
        close = test_data['close']


        stream = ta_indicators.ApoStream(short_period=10, long_period=20)


        stream_results = []
        for price in close:
            result = stream.update(price)
            stream_results.append(result)


        stream_results = [np.nan if x is None else x for x in stream_results]


        batch_results = ta_indicators.apo(close, short_period=10, long_period=20)


        assert_close(stream_results, batch_results, rtol=1e-10)

    def test_apo_batch_single_parameter_set(self, test_data):
        """Test batch processing with single parameter combination"""
        close = test_data['close']


        batch_result = ta_indicators.apo_batch(
            close,
            short_period_range=(10, 10, 0),
            long_period_range=(20, 20, 0)
        )


        assert 'values' in batch_result
        assert 'short_periods' in batch_result
        assert 'long_periods' in batch_result


        assert batch_result['values'].shape == (1, len(close))
        assert len(batch_result['short_periods']) == 1
        assert len(batch_result['long_periods']) == 1


        single_result = ta_indicators.apo(close, short_period=10, long_period=20)
        assert_close(batch_result['values'][0], single_result, rtol=1e-10)

    def test_apo_batch_multiple_parameters(self, test_data):
        """Test batch processing with multiple parameter combinations"""
        close = test_data['close'][:100]


        batch_result = ta_indicators.apo_batch(
            close,
            short_period_range=(5, 15, 5),
            long_period_range=(20, 30, 10)
        )


        expected_combos = [
            (5, 20), (5, 30),
            (10, 20), (10, 30),
            (15, 20), (15, 30)
        ]

        assert batch_result['values'].shape[0] == len(expected_combos)
        assert batch_result['values'].shape[1] == len(close)
        assert len(batch_result['short_periods']) == len(expected_combos)
        assert len(batch_result['long_periods']) == len(expected_combos)


        for i, (short, long) in enumerate(expected_combos):
            single_result = ta_indicators.apo(close, short_period=short, long_period=long)
            assert_close(
                batch_result['values'][i],
                single_result,
                rtol=1e-10,
                msg=f"Batch row {i} (short={short}, long={long}) mismatch"
            )

    def test_apo_batch_invalid_combinations(self, test_data):
        """Test batch processing filters out invalid combinations"""
        close = test_data['close'][:50]


        batch_result = ta_indicators.apo_batch(
            close,
            short_period_range=(10, 30, 10),
            long_period_range=(15, 25, 10)
        )




        assert batch_result['values'].shape[0] == 3


        params = list(zip(batch_result['short_periods'], batch_result['long_periods']))
        assert (10, 15) in params
        assert (10, 25) in params
        assert (20, 25) in params

    def test_apo_batch_empty_result(self):
        """Test batch processing with no valid combinations"""
        close = np.random.rand(50)


        with pytest.raises(ValueError):
            ta_indicators.apo_batch(
                close,
                short_period_range=(20, 30, 10),
                long_period_range=(10, 15, 5)
            )

    def test_apo_edge_cases(self):
        """Test APO with edge case inputs"""

        min_data = np.array([1.0] * 20)
        result = ta_indicators.apo(min_data, short_period=10, long_period=20)
        assert len(result) == len(min_data)


        assert result[0] == 0.0


        assert abs(result[-1]) < 1e-10